tests/test_type_adapter.py

```diff
@@ -309,3 +308,45 @@ def test_validate_strings_dict(strict):
         1: date(2017, 1, 1),
         2: date(2017, 1, 2),
     }
+
+
+def test_validate_type_adapter_config_error() -> None:
+    class Model1(BaseModel):
+        x: int
+
+    @dataclass
+    class Model2:
+        x: int
+
+        __pydantic_config__ = ConfigDict(strict=False)
+
+    class Model3(TypedDict):
+        x: int
+
+        __pydantic_config__ = ConfigDict(strict=False)
+
+    for model in (Model1, Model2, Model3):
+        with pytest.raises(
+            PydanticUserError,
+            match='Cannot use `config` when the adapted type is a BaseModel or has `__pydantic_config__` defined.',
+        ):
+            TypeAdapter(model, config=ConfigDict(strict=False))
+
+
+def test_type_adapter_config_nested() -> None:
+    class ExampleNested(TypedDict):
+        x: Union[str, 'ExampleNested']
+        y: str
+
+    ta = TypeAdapter(ExampleNested, config=ConfigDict(str_to_upper=True))
+    assert ta.validate_python({'x': 'foo', 'y': 'bar'}) == {'x': 'FOO', 'y': 'BAR'}
+    assert ta.validate_json('{"x": "foo", "y": "bar"}') == {'x': 'FOO', 'y': 'BAR'}
+
+    assert ta.validate_python({'x': {'x': 'foo', 'y': 'baz'}, 'y': 'bar'}) == {
+        'x': {'x': 'FOO', 'y': 'BAZ'},
```

Comment:
Is this right or wrong?

I think we may need an even more complex example: put something _with_ a config in the middle of the recursion. So for example:


```python
class Model(BaseModel):
    inner: Nested

class Nested(TypedDict):
    x: Model | None
    y: str
```

Then I believe `{'x': {'inner': {'x': None, 'y': 'a'}}, 'y': 'a'}` should validate to `{'x': {'inner': {'x': None, 'y': 'a'}}, 'y': 'A'}` since the config on `Model` prevents any pushdown.
In other words, if the result in the test came from pushdown that'd be fine but it's not, it's coming from attaching the config to the ref even though the config is not part of the ref's identifier.

If we could incorporate the config into the ref as an identifier then we'd end up with two different ref's and all would be good. But we can't, hence why we disallow it completely.

Reply:
@adriangb, these are the results I get from such a test:

```py
def test_type_adapter_config_tested_with_model() -> None:
    class Model(BaseModel):
        inner: 'Nested'

    class Nested(TypedDict):
        x: Union[Model, None]
        y: str

    ta = TypeAdapter(Nested, config=ConfigDict(str_to_upper=True))
    assert ta.validate_python({'x': {'inner': {'x': None, 'y': 'a'}}, 'y': 'a'}) == {
        'x': Model(inner={'x': None, 'y': 'A'}),  # <-- note, the cap A
        'y': 'A'
    }
```

Reply:
Indeed that’s problematic

Reply:
I think the only way to make it work is to add a `with_config` validator that puts it into ValidationState

Reply:
@adriangb, would the `with_config` validator go directly on the `Nested` model in the use case above?

Reply:
I mean that TypeAdapter would then blindly wrap the schema it generates with that

Reply:
Ah, I thought that we had a runtime `CoreConfig` being passed around. We unfortunately don't. So I don't think that approach would be possible unless we fundamentally changed how `Config` gets passed around in `pydantic-core` :(

Reply:
Ah, darn. Good to know, thanks. Excited to chat more on Monday.